Fixing inertia calcs

BrightKnight02 opened the pull request:
Trying to fix the inertia calculations for the optimization. The was a small error in the slender rod calculations, which was found after creating the tests in `test_inertial_preperties.py`. 

I'm not going to close the issue yet as I may be interpreting some things wrong.

There is a chance things that were not supposed to change did change. For those things I am sorry.

darint916 committed:
Merge pull request #18 from darint916/main

Moving inertia tensors to be about wing center of mass

BrightKnight02 committed:
Fixing me accidentally deleting changes
I am bad at git

BrightKnight02 committed:
Fixed slender rod inertia calcs
FOund the math error causing the inertia tensor of a slender rod to be calculated wrong

BrightKnight02 committed:
Update test_inertial_preperties.py

math notation

darint916 committed:
Update test_wing_classes.py spacing

darint916 commented on flapping_proj/unit_tests/test_inertial_preperties.py:
Write failing asserts? for the unit test library to catch failures

darint916 commented on flapping_proj/unit_tests/test_inertial_preperties.py:
```diff
@@ -0,0 +1,114 @@
+            calc_com = calc_com / self.truth_mass[i]
+            Message.info(f"Analysing rod with endpoints (y, z) at ({coords[0]}, {coords[1]}) and ({coords[2]}, {coords[3]})")
+            
+            if almost_equal(truth_com[i][0], calc_com[0], self.error_tol) and almost_equal(truth_com[i][1], calc_com[1], self.error_tol):
```
for assert do:
self.assertTrue(almost_equal(truth_com[i][0], calc_com[0], self.error_tol) and almost_equal(truth_com[i][1], calc_com[1], self.error_tol, Message.error("Failed case"))

BrightKnight02 replied:
Updated with https://github.com/darint916/ros2-gazebo-drake-sims/pull/19/commits/7cca65e214853907c57dae233d516940d168f046

darint916 reviewed (changes requested):
small changes in formatting and fixing testing

BrightKnight02 committed:
Merge branch 'main' into fixing_inertia_calcs

BrightKnight02 committed:
Changed inertia tests to assert statements rather than Message classes

Co-authored-by: Darin  o - o <darint916@users.noreply.github.com>

darint916 committed:
rename fix for smoke test setup on top level run.
Bug still found in tensor calculation

darint916 commented:
Test cases passed unit tests, but failed on top-level smoke test, new test case was made with given opt parameters:
0.00484565 -0.00135816  0.03554422 -0.06507584  0.09359727 -0.00165281 0.51404022
produced values of 
'ixx': 3.392129987389574e-06, 'ixy': 0.0, 'ixz': 0.0, 'iyy': 1.9036163412645447e-06, 'iyz': -2.629275870389845e-06, 'izz': 1.4885136461250288e-06

Solidworks truth values as:
Moments of inertia: ( grams *  square millimeters )
Taken at the center of mass and aligned with the output coordinate system. (Using positive tensor notation.)
    Lxx = 197.457736    Lxy = 0.000000    Lxz = 0.000000
    Lyx = 0.000000    Lyy = 73.398949    Lyz = 23.131955
    Lzx = 0.000000    Lzy = 23.131955    Lzz = 124.076478

darint916 reviewed (changes requested):
Test cases passed unit tests, but failed on top-level smoke test, new test case was made with given opt parameters:
`0.00484565 -0.00135816  0.03554422 -0.06507584  0.09359727 -0.00165281 0.51404022`
produced values of 
`'ixx': 3.392129987389574e-06, 'ixy': 0.0, 'ixz': 0.0, 'iyy': 1.9036163412645447e-06, 'iyz': -2.629275870389845e-06, 'izz': 1.4885136461250288e-06`

Solidworks truth values as:
Moments of inertia: ( grams *  square millimeters )
Taken at the center of mass and aligned with the output coordinate system. (Using positive tensor notation.)
 `   Lxx = 197.457736    Lxy = 0.000000    Lxz = 0.000000
    Lyx = 0.000000    Lyy = 73.398949    Lyz = 23.131955
    Lzx = 0.000000    Lzy = 23.131955    Lzz = 124.076478`
    
    
    All of Inertia Properties of Rod , COM, total mass, moment of inertia tested.
    
    Seems to be failing on the film.

BrightKnight02 committed:
Started film test writing
Established the file and some basic test cases for the film calculations. Potentially found a math error while trying to write more interesting cases, but am not fixing with this commit.

BrightKnight02 committed:
Mostly fixed film inertia
Film inertia functions are passing all tests except the center of mass test for Para1. From what I can tell, scipy is just evaluating the integral incorrectly.

BrightKnight02 committed:
Film calculations tested and succeeds in all tests
No actual error was found in the center of mass calculation, the truth value was just calculated incorrectly. Film calculators pass all tests.

BrightKnight02 commented:
This is ready for an additional review. https://github.com/darint916/ros2-gazebo-drake-sims/pull/19/commits/d1ec297ed6a17289ca889a7ec48cf7fd150428fd should have fixed all calculations related to the current optimization.

BrightKnight02 committed:
Adding a massive testset to `test_wing_classes`
This commit incluides a for loop abomination that loops over 10000 valid wings to see if the generated inertia tensor is valid. At the moment, there is a failure at a wing defined by the points (0.03, 0), (0.1485, -0.2), (0.15, 0). What I am likely going to try for the next commit is to retrofit all the inertia calculations to calculate relative to the center of mass

BrightKnight02 committed:
Calculate inertia about center of mass
It's finally been done the correct way. All tests pass.

darint916 committed:
Fixed Inertia calculations, added com for inertial, swapped x and y reference axis.

Co-authored-by: BrightKnight <BrightKnight02@users.noreply.github.com>

BrightKnight02 commented:
The inertia was changed to be in the correct coordinate system. Inertia and center of mass was calculated as if the wing spar was along the y axis, but Gazebo was set up such that the wing spar was along the x axis. This was fixed by changing what index lines 83-88 and line 92 in `top_level.py` pointed to for the inertia tensor calculated by the `TriWing` class. This is equivalent to rotating the coordinate system.
